[Tensorpipe Agent] Enabling tests with OSS CI

osalpekar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#38447 [Tensorpipe Agent] Enabling tests with OSS CI**
* #38446 [Tensorpipe Agent] Dist Optimizer Tests for Tensorpipe Agent
* #38445 [Tensorpipe Agent] Dist Autograd Tests for Tensorpipe Agent
* #38444 [Tensorpipe Agent] RPC, RRef tests for Tensorpipe Agent

This PR modifies `run_tests.py` to enable running Tensorpipe Agent tests with the OSS CI.

Fixes: https://github.com/pytorch/pytorch/issues/36075

Differential Revision: [D21560096](https://our.internmc.facebook.com/intern/diff/D21560096/)

osalpekar committed:
[Tensorpipe Agent] Enabling tests with OSS CI

This PR modifies `run_tests.py` to enable running Tensorpipe Agent tests with the OSS CI.

Differential Revision: [D21560096](https://our.internmc.facebook.com/intern/diff/D21560096/)

[ghstack-poisoned]

lw commented:
From the CI it seems there may be a few more tests that we need to disable for now:
```
May 13 15:46:34 ======================================================================
May 13 15:46:34 FAIL [1.585s]: test_nested_rref_stress (__main__.TensorPipeAgentRpcTestWithSpawn)
May 13 15:46:34 ----------------------------------------------------------------------
May 13 15:46:34 Traceback (most recent call last):
May 13 15:46:34   File "/Users/distiller/workspace/miniconda3/lib/python3.7/site-packages/torch/testing/_internal/common_distributed.py", line 214, in wrapper
May 13 15:46:34     self._join_processes(fn)
May 13 15:46:34   File "/Users/distiller/workspace/miniconda3/lib/python3.7/site-packages/torch/testing/_internal/common_distributed.py", line 316, in _join_processes
May 13 15:46:34     self._check_return_codes(elapsed_time)
May 13 15:46:34   File "/Users/distiller/workspace/miniconda3/lib/python3.7/site-packages/torch/testing/_internal/common_distributed.py", line 359, in _check_return_codes
May 13 15:46:34     i, first_process.exitcode, p.exitcode
May 13 15:46:34   File "/Users/distiller/workspace/miniconda3/lib/python3.7/site-packages/torch/testing/_internal/common_utils.py", line 1010, in assertEqual
May 13 15:46:34     super().assertLessEqual(abs(x - y), atol, message)
May 13 15:46:34 AssertionError: 6 not less than or equal to 1e-05 : Expect process 2 exit code to match Process 0 exit code of 0, but got -6
May 13 15:46:34 
May 13 15:46:34 ----------------------------------------------------------------------
```
```
May 13 22:55:22 ======================================================================
May 13 22:55:22 FAIL [1.620s]: test_py_rref_args_user_share (__main__.TensorPipeAgentRpcTestWithSpawn)
May 13 22:55:22 ----------------------------------------------------------------------
May 13 22:55:22 Traceback (most recent call last):
May 13 22:55:22   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 214, in wrapper
May 13 22:55:22     self._join_processes(fn)
May 13 22:55:22   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 316, in _join_processes
May 13 22:55:22     self._check_return_codes(elapsed_time)
May 13 22:55:22   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 359, in _check_return_codes
May 13 22:55:22     i, first_process.exitcode, p.exitcode
May 13 22:55:22   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 1010, in assertEqual
May 13 22:55:22     super().assertLessEqual(abs(x - y), atol, message)
May 13 22:55:22 AssertionError: 6 not less than or equal to 1e-05 : Expect process 1 exit code to match Process 0 exit code of 0, but got -6
May 13 22:55:22 
May 13 22:55:22 ----------------------------------------------------------------------
```
```
May 14 01:25:52   test_cuda (__main__.TensorPipeAgentRpcTestWithSpawn) ... Timing out after 100 seconds and killing subprocesses.
May 14 01:25:52 ERROR (100.101s)
```

osalpekar commented:
> From the CI it seems there may be a few more tests that we need to disable for now:
> 
> ```
> May 13 15:46:34 ======================================================================
> May 13 15:46:34 FAIL [1.585s]: test_nested_rref_stress (__main__.TensorPipeAgentRpcTestWithSpawn)
> May 13 15:46:34 ----------------------------------------------------------------------
> May 13 15:46:34 Traceback (most recent call last):
> May 13 15:46:34   File "/Users/distiller/workspace/miniconda3/lib/python3.7/site-packages/torch/testing/_internal/common_distributed.py", line 214, in wrapper
> May 13 15:46:34     self._join_processes(fn)
> May 13 15:46:34   File "/Users/distiller/workspace/miniconda3/lib/python3.7/site-packages/torch/testing/_internal/common_distributed.py", line 316, in _join_processes
> May 13 15:46:34     self._check_return_codes(elapsed_time)
> May 13 15:46:34   File "/Users/distiller/workspace/miniconda3/lib/python3.7/site-packages/torch/testing/_internal/common_distributed.py", line 359, in _check_return_codes
> May 13 15:46:34     i, first_process.exitcode, p.exitcode
> May 13 15:46:34   File "/Users/distiller/workspace/miniconda3/lib/python3.7/site-packages/torch/testing/_internal/common_utils.py", line 1010, in assertEqual
> May 13 15:46:34     super().assertLessEqual(abs(x - y), atol, message)
> May 13 15:46:34 AssertionError: 6 not less than or equal to 1e-05 : Expect process 2 exit code to match Process 0 exit code of 0, but got -6
> May 13 15:46:34 
> May 13 15:46:34 ----------------------------------------------------------------------
> ```
> 
> ```
> May 13 22:55:22 ======================================================================
> May 13 22:55:22 FAIL [1.620s]: test_py_rref_args_user_share (__main__.TensorPipeAgentRpcTestWithSpawn)
> May 13 22:55:22 ----------------------------------------------------------------------
> May 13 22:55:22 Traceback (most recent call last):
> May 13 22:55:22   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 214, in wrapper
> May 13 22:55:22     self._join_processes(fn)
> May 13 22:55:22   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 316, in _join_processes
> May 13 22:55:22     self._check_return_codes(elapsed_time)
> May 13 22:55:22   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 359, in _check_return_codes
> May 13 22:55:22     i, first_process.exitcode, p.exitcode
> May 13 22:55:22   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 1010, in assertEqual
> May 13 22:55:22     super().assertLessEqual(abs(x - y), atol, message)
> May 13 22:55:22 AssertionError: 6 not less than or equal to 1e-05 : Expect process 1 exit code to match Process 0 exit code of 0, but got -6
> May 13 22:55:22 
> May 13 22:55:22 ----------------------------------------------------------------------
> ```
> 
> ```
> May 14 01:25:52   test_cuda (__main__.TensorPipeAgentRpcTestWithSpawn) ... Timing out after 100 seconds and killing subprocesses.
> May 14 01:25:52 ERROR (100.101s)
> ```

Skipped these in the first diff in the stack, but there are sure be more failures due to flaky tests

osalpekar committed:
Update on "[Tensorpipe Agent] Enabling tests with OSS CI"


This PR modifies `run_tests.py` to enable running Tensorpipe Agent tests with the OSS CI.

Differential Revision: [D21560096](https://our.internmc.facebook.com/intern/diff/D21560096/)

[ghstack-poisoned]

lw commented:
I'm still seeing errors:
```
May 14 23:38:35 ======================================================================
May 14 23:38:35 FAIL [1.422s]: test_remote_same_worker (__main__.TensorPipeAgentRpcTestWithSpawn)
May 14 23:38:35 ----------------------------------------------------------------------
May 14 23:38:35 Traceback (most recent call last):
May 14 23:38:35   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 204, in wrapper
May 14 23:38:35     self._join_processes(fn)
May 14 23:38:35   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 306, in _join_processes
May 14 23:38:35     self._check_return_codes(elapsed_time)
May 14 23:38:35   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 349, in _check_return_codes
May 14 23:38:35     i, first_process.exitcode, p.exitcode
May 14 23:38:35   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 1037, in assertEqual
May 14 23:38:35     super().assertLessEqual(abs(x - y), atol, message)
May 14 23:38:35 AssertionError: 6 not less than or equal to 1e-05 : Expect process 1 exit code to match Process 0 exit code of -6, but got 0
May 14 23:38:35 
May 14 23:38:35 ----------------------------------------------------------------------
```
```
May 14 23:35:27 ======================================================================
May 14 23:35:27 FAIL [1.323s]: test_use_rref_after_shutdown (__main__.TensorPipeAgentRpcTestWithSpawn)
May 14 23:35:27 ----------------------------------------------------------------------
May 14 23:35:27 Traceback (most recent call last):
May 14 23:35:27   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 204, in wrapper
May 14 23:35:27     self._join_processes(fn)
May 14 23:35:27   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 306, in _join_processes
May 14 23:35:27     self._check_return_codes(elapsed_time)
May 14 23:35:27   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 349, in _check_return_codes
May 14 23:35:27     i, first_process.exitcode, p.exitcode
May 14 23:35:27   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 1037, in assertEqual
May 14 23:35:27     super().assertLessEqual(abs(x - y), atol, message)
May 14 23:35:27 AssertionError: 6 not less than or equal to 1e-05 : Expect process 2 exit code to match Process 0 exit code of 0, but got -6
May 14 23:35:27 
May 14 23:35:27 ----------------------------------------------------------------------
```
and the `test_cuda` one I mentioned yesterday.

They could very well just be flakes, but this is clearly a flakiness that only happens with the TensorPipe agent. The error message is admittedly a bit cryptic, but it says that one process exited with SIGABORT, which is what happens in case of an uncaught exception. Unfortunately there's no mention of what exception that is nor of a backtrace. In any case, this is related to error handling. I'd be inclined to go the easy route and just disable all the flaky tests for now, and then after fixing the error handling we can reenable them after making sure they're not flaky anymore.

osalpekar committed:
Update on "[Tensorpipe Agent] Enabling tests with OSS CI"


This PR modifies `run_tests.py` to enable running Tensorpipe Agent tests with the OSS CI.

Differential Revision: [D21560096](https://our.internmc.facebook.com/intern/diff/D21560096/)

[ghstack-poisoned]

osalpekar committed:
Update on "[Tensorpipe Agent] Enabling tests with OSS CI"


This PR modifies `run_tests.py` to enable running Tensorpipe Agent tests with the OSS CI.

Fixes: https://github.com/pytorch/pytorch/issues/36075

Differential Revision: [D21560096](https://our.internmc.facebook.com/intern/diff/D21560096/)

[ghstack-poisoned]

osalpekar commented:
I've stress-tested each test 100x and skipped everything that fails even once (including the ones linked above). This should be pretty comprehensive, but it is still completely possible that more tests fail in the CI/Sandcastle (whether the test is flaky with TensorPipe specifically, or just in general). For example, `test_callback_simple` is flaky across all agents with both fork and spawn. I think we should be in pretty good shape to land this, but if there is flakiness observed afterwards, we can simply disable those tests as we do with ProcessGroup tests.

mrshenli commented on test/run_test.py:
```diff
@@ -73,6 +73,9 @@
     'distributed/rpc/faulty_agent/test_dist_autograd_spawn',
     'distributed/rpc/faulty_agent/test_rpc_spawn',
     'distributed/rpc/jit/test_dist_autograd_spawn',
+    'distributed/rpc/tensorpipe/test_dist_autograd_spawn',
```
Shall we also add these to `SLOW_TESTS`?

mrshenli reviewed (changes requested):
Oops, test failures are real:

```
May 16 00:44:28 ======================================================================
May 16 00:44:28 FAIL [1.625s]: test_rref_get_future (__main__.TensorPipeAgentRpcTestWithSpawn)
May 16 00:44:28 ----------------------------------------------------------------------
May 16 00:44:28 Traceback (most recent call last):
May 16 00:44:28   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 204, in wrapper
May 16 00:44:28     self._join_processes(fn)
May 16 00:44:28   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 306, in _join_processes
May 16 00:44:28     self._check_return_codes(elapsed_time)
May 16 00:44:28   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 349, in _check_return_codes
May 16 00:44:28     i, first_process.exitcode, p.exitcode
May 16 00:44:28   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 1037, in assertEqual
May 16 00:44:28     super().assertLessEqual(abs(x - y), atol, message)
May 16 00:44:28 AssertionError: 6 not less than or equal to 1e-05 : Expect process 1 exit code to match Process 0 exit code of -6, but got 0
May 16 00:44:28 
```

osalpekar committed:
Update on "[Tensorpipe Agent] Enabling tests with OSS CI"


This PR modifies `run_tests.py` to enable running Tensorpipe Agent tests with the OSS CI.

Fixes: https://github.com/pytorch/pytorch/issues/36075

Differential Revision: [D21560096](https://our.internmc.facebook.com/intern/diff/D21560096/)

[ghstack-poisoned]

osalpekar committed:
Update on "[Tensorpipe Agent] Enabling tests with OSS CI"


This PR modifies `run_tests.py` to enable running Tensorpipe Agent tests with the OSS CI.

Fixes: https://github.com/pytorch/pytorch/issues/36075

Differential Revision: [D21560096](https://our.internmc.facebook.com/intern/diff/D21560096/)

[ghstack-poisoned]

osalpekar committed:
Update on "[Tensorpipe Agent] Enabling tests with OSS CI"


This PR modifies `run_tests.py` to enable running Tensorpipe Agent tests with the OSS CI.

Fixes: https://github.com/pytorch/pytorch/issues/36075

Differential Revision: [D21560096](https://our.internmc.facebook.com/intern/diff/D21560096/)

[ghstack-poisoned]

lw commented:
There still seem to be a few tests failing...

osalpekar committed:
Update on "[Tensorpipe Agent] Enabling tests with OSS CI"


This PR modifies `run_tests.py` to enable running Tensorpipe Agent tests with the OSS CI.

Fixes: https://github.com/pytorch/pytorch/issues/36075

Differential Revision: [D21560096](https://our.internmc.facebook.com/intern/diff/D21560096/)

[ghstack-poisoned]

osalpekar committed:
Update on "[Tensorpipe Agent] Enabling tests with OSS CI"


This PR modifies `run_tests.py` to enable running Tensorpipe Agent tests with the OSS CI.

Fixes: https://github.com/pytorch/pytorch/issues/36075

Differential Revision: [D21560096](https://our.internmc.facebook.com/intern/diff/D21560096/)

[ghstack-poisoned]

osalpekar committed:
Update on "[Tensorpipe Agent] Enabling tests with OSS CI"


This PR modifies `run_tests.py` to enable running Tensorpipe Agent tests with the OSS CI.

Fixes: https://github.com/pytorch/pytorch/issues/36075

Differential Revision: [D21560096](https://our.internmc.facebook.com/intern/diff/D21560096/)

[ghstack-poisoned]

osalpekar commented:
Failure (test_nccl) looks unrelated. All other tests seem to be passing/skipped

jiayisuse commented:
Looks good to me.  The NCCL test failure seems not related?
```
01:25:18 ======================================================================
01:25:18 FAIL: test_broadcast (__main__.TestNCCL)
01:25:18 ----------------------------------------------------------------------
01:25:18 Traceback (most recent call last):
01:25:18   File "distributed/test_nccl.py", line 39, in test_broadcast
01:25:18     self.assertEqual(tensors[i], expected)
01:25:18   File "/var/lib/jenkins/.local/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 1083, in assertEqual
01:25:18     self.assertTrue(result, msg=message)
01:25:18 AssertionError: False is not true : Tensors failed to compare as equal! With rtol=1.3e-06 and atol=1e-05, found 128 element(s) (out of 128) whose difference(s) exceeded the margin of error (including 0 nan comparisons). The greatest difference was 5.465000152587891 (5.587951183319092 vs. 0.12295103073120117), which occurred at index 84.
01:25:18 
01:25:18 ======================================================================
01:25:18 FAIL: test_reduce (__main__.TestNCCL)
01:25:18 ----------------------------------------------------------------------
01:25:18 Traceback (most recent call last):
01:25:18   File "distributed/test_nccl.py", line 52, in test_reduce
01:25:18     self.assertEqual(tensors[0], expected)
01:25:18   File "/var/lib/jenkins/.local/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 1083, in assertEqual
01:25:18     self.assertTrue(result, msg=message)
01:25:18 AssertionError: False is not true : Tensors failed to compare as equal! With rtol=1.3e-06 and atol=1e-05, found 128 element(s) (out of 128) whose difference(s) exceeded the margin of error (including 0 nan comparisons). The greatest difference was 5.54068660736084 (11.835615158081055 vs. 6.294928550720215), which occurred at index 105.
01:25:18 
01:25:18 ======================================================================
01:25:18 FAIL: test_reduce_scatter (__main__.TestNCCL)
01:25:18 ----------------------------------------------------------------------
01:25:18 Traceback (most recent call last):
01:25:18   File "distributed/test_nccl.py", line 100, in test_reduce_scatter
01:25:18     self.assertEqual(outputs[i], expected[i])
01:25:18   File "/var/lib/jenkins/.local/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 1083, in assertEqual
01:25:18     self.assertTrue(result, msg=message)
01:25:18 AssertionError: False is not true : Tensors failed to compare as equal! With rtol=1.3e-06 and atol=1e-05, found 32 element(s) (out of 32) whose difference(s) exceeded the margin of error (including 0 nan comparisons). The greatest difference was 9.568137645721436 (14.023767471313477 vs. 4.455629825592041), which occurred at index 30.
```

osalpekar commented:
@jiayisuse Yeah it's unrelated, `test_nccl` isn't an RPC/Dist Autograd test

mrshenli commented:
This has caused too many test failures on master. See #38758 #38757 #38756 #38755. I wonder if we should revert/disable this for now, use those linked CI env in the issues to debug? We can add this back when those tests are less flaky.

@osalpekar @jiayisuse @lw @mruberry

mrshenli commented:
Sorry that I have to revert this, test failures keep popping on master that shadows real test signals.  Let me know if you need more info on how to debug in CI envs.

lw commented:
No worries, I'm already debugging these tests and hopefully we can re-land it with everything enabled and passing soon!
